6941 Number Components Rounding

aimee-mcneil-melville opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Part 2 of #6941

# 👩🏻‍💻 What does this PR do?

Add functionality to common components to make them more reuseable for numbers display

Reuseable tooltip function to localise numbers and round the tooltip display to 10 decimal places

NumericTextInput - has a tooltip to 10 dp when disabled
NumericTextDisplay - formats the number nicely for display, not a cell or input
NumberCell - uses NumericTextDisplay for formatting the number

Almost all numbers are required to be in 2dp with a tooltip up to 10dp except DOSES which should all be whole numbers. will need to follow up and undo some of the previous PR for this issue

Prescription -> Line edit view
<img width="206" height="178" alt="Screenshot 2025-07-22 at 9 51 18 AM" src="https://github.com/user-attachments/assets/94263f1a-2044-430e-a0a8-7fd44d1419c5" />

Internal Orders
<img width="1217" height="814" alt="Screenshot 2025-07-22 at 6 57 48 AM" src="https://github.com/user-attachments/assets/f3fc0758-e2db-4cf4-bac8-684323cb445c" />

Requisition
<img width="1217" height="814" alt="Screenshot 2025-07-22 at 6 51 44 AM" src="https://github.com/user-attachments/assets/769fbe7f-571d-4943-9ab9-e2d7a9cc0ebf" />

Item
<img width="1212" height="594" alt="Screenshot 2025-07-22 at 6 30 37 AM" src="https://github.com/user-attachments/assets/74fbf3c0-b8e5-4dfd-88db-25f95c00518b" />

View Stock -> Repack
<img width="913" height="716" alt="Screenshot 2025-07-22 at 6 00 21 AM" src="https://github.com/user-attachments/assets/1447c42f-888c-44dc-bc5d-0d7bb1f3ad03" />

View Stock -> Adjust
<img width="727" height="604" alt="Screenshot 2025-07-22 at 8 22 40 AM" src="https://github.com/user-attachments/assets/db51300c-0ddc-4624-9557-6886ac12b278" />
<!-- Explain the changes you made -->

** PR Review additions

Doses: Doses should always show as whole units.

I looked at using useMemo and useFormatNumber in the [utils](https://github.com/msupply-foundation/open-msupply/blob/6941-rounding-of-figures/client/packages/requisitions/src/common/utils.ts#L17) for dose value, but React rules.... added the comment instead and used rounding in the components. EG:

Requisitions line view:
<img width="412" height="333" alt="Screenshot 2025-07-23 at 1 36 53 PM" src="https://github.com/user-attachments/assets/018151e8-06ec-4a71-ae20-4652ad3fe0f9" />

Internal orders line view:
<img width="596" height="355" alt="Screenshot 2025-07-23 at 1 36 05 PM" src="https://github.com/user-attachments/assets/e3f1f49a-3406-4c5e-830c-baef837a0a60" />

Difference Column -> Stocktake Detail View:
<img width="458" height="279" alt="Screenshot 2025-07-23 at 12 47 56 PM" src="https://github.com/user-attachments/assets/27e136df-5a12-4729-a8a7-396300a2de18" />

useClosedSummary - Prescription Line Edit View
<img width="927" height="179" alt="Screenshot 2025-07-23 at 1 28 38 PM" src="https://github.com/user-attachments/assets/b977251b-2ebd-4425-9a41-ede225723a36" />
iption Line Edit View

DoseQtyColumn - added a helper so the value can be rounded with useFormatNumber. Example usage in:
Prescription Detail View
<img width="465" height="254" alt="Screenshot 2025-07-23 at 1 35 23 PM" src="https://github.com/user-attachments/assets/60e0e05f-c859-4b7c-805d-831f0ce5c817" />

Outbound Shipment Detail View:
<img width="338" height="279" alt="Screenshot 2025-07-23 at 1 20 33 PM" src="https://github.com/user-attachments/assets/29a4ed1b-2e72-4115-8775-0b0edf13d61e" />


<!-- why are the changes needed -->

<!-- Add a screenshot if there are UI changes  -->

## 💌 Any notes for the reviewer?

I must have branched off another branch here - extra commits 😓 The are merged into the same base though so not showing in the diff

<!-- Do you have any specific questions for the reviewer? -->

<!-- Is there a high risk/complicated change they should focus on? -->

<!-- any general areas of the codebase touched? any side effects caused? -->

<!-- Anything half cooked but going to be finished off in a different PR? -->

# 🧪 Testing

<!-- Explain the steps you'd take to test the changes of this PR manually -->

- [ ]  Have a vaccine item with a number of doses and units per pack that will create a long decimals if one dose is given eg. I used 3 doses per unit, 7 units per pack
- [ ]  Prescribe one dose of the vaccine
- [ ]  Use this item where changes have been made, check for 2 decimal place rounding and a tooltip to 10 decimal place
- [ ]  Change language -> see where regional settings have , or . as the separator, the tooltip matches
- [ ] See doses always displayed as whole numbers


# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: <!-- _(e.g.)_ New `issued` column in `Requisitions` indicates stock quantity already in shipments -->
  1.
  2.


# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

aimee-mcneil-melville commented on client/packages/common/src/intl/number/IntlNumber.ts:
```diff
@@ -43,6 +43,16 @@ export const useFormatNumber = () => {
       });
       return intl.format(value ?? 0);
     },
+    tooltip: (value?: number, dp?: number): string => {
```
New common function to handle tooltip formatting

aimee-mcneil-melville replied:
Yes it is! My thought process for this was reusability and consistency - if for any reason we want to change how many DP shows by default in the tooltips we can change that in just one place, and default to 10 instead of 0 if no DP is passed in

aimee-mcneil-melville commented on client/packages/invoices/src/Prescriptions/DetailView/columns.ts:
```diff
@@ -51,10 +50,7 @@ export const useExpansionColumns = (
-            (rowData.numberOfPacks ?? 0) * (rowData.packSize ?? 1),
-            3
-          ),
+          (rowData.numberOfPacks ?? 0) * (rowData.packSize ?? 1),
```
Default to 2dp rounding for prescriptions

aimee-mcneil-melville commented on client/packages/system/src/Item/Components/PackSize/PackQuantityCell.tsx:
```diff
@@ -3,31 +3,20 @@ import {
-            : roundedPackQuantity}
-        </Typography>
-      </Tooltip>
+      <NumericTextDisplay value={packQuantity} />
```
Again reused the component here for rendering the cell

aimee-mcneil-melville commented on client/packages/system/src/Stock/Components/InventoryAdjustment/InventoryAdjustmentModal.tsx:
```diff
@@ -154,14 +151,12 @@ export const InventoryAdjustmentModal = ({
-                      : -draft.adjustment)
-                  ).toFixed(2)
-                )}
+                value={
```
👀 I can't replicate this.. There is a `decimalLimit={2}` already on the input field, and I can only enter 2dp there. The new quantity is always disabled and only displaying the calculated value

roxy-dao replied:
Oh strange... it works now 👀

aimee-mcneil-melville commented:
Thanks for all the feedback Roxy! 
I have removed the tooltip function and changed all dose implementations to be rounded to 0 decimal places (some introduced [here](https://github.com/msupply-foundation/open-msupply/pull/8429)).

For the ellipses I did notice that too - it's part of the underlying floating point calculations and I've raised some further discussion about how we handle numbers close to X precision

aimee-mcneil-melville commented on client/packages/requisitions/src/common/utils.test.ts:
```diff
@@ -68,16 +68,4 @@ describe('calculateValueInDoses', () => {
     // 2.5 units × 2 doses per unit = 5 doses
     expect(calculateValueInDoses(Representation.UNITS, 10, 2, 2.5)).toBe(5);
   });
-
```
I removed this test as the function does no rounding, but tests still passed with `toBeCloseTo` 🫠

roxy-dao replied:
Yeah it's a matching thing was accounting for if it does 33.333333332
